File: src/main/java/com/example/cqrs/entity/read/write/event/MoneyWithdrawnEvent.java
package com.example.cqrs.entity.read.write.event;

import event_sourcing.study.entity.write.event.base.BaseAccountEvent;
import event_sourcing.study.entity.write.event.base.EventMetadata;
import jakarta.persistence.DiscriminatorValue;
import jakarta.persistence.Entity;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@NoArgsConstructor(access = AccessLevel.PROTECTED)
@DiscriminatorValue("MoneyWithdrawnEvent")
@Entity
public class MoneyWithdrawnEvent extends BaseAccountEvent {

    public MoneyWithdrawnEvent(String accountId, LocalDateTime eventDate,
                              double amount, EventMetadata metadata) {
        super(accountId, eventDate, amount, metadata);
    }

}

File: src/main/java/com/example/cqrs/infrastructure/eventstore/entity/event/money/MoneyWithdrawnEvent.kt
package com.example.cqrs.infrastructure.eventstore.entity.event.money

import com.example.cqrs.infrastructure.eventstore.entity.base.AccountEvent
import com.example.cqrs.infrastructure.eventstore.entity.base.metadata.EventMetadata
import jakarta.persistence.DiscriminatorValue
import jakarta.persistence.Entity
import java.time.LocalDateTime

/**
 * 출금 이벤트
 */
@DiscriminatorValue("MoneyWithdrawnEvent")
@Entity
class MoneyWithdrawnEvent(
    accountId: String,
    amount: Double,
    eventDate: LocalDateTime,
    metadata: EventMetadata
) : AccountEvent(
    id = null,
    accountId = accountId,
    amount = amount,
    eventDate = eventDate,
    metadata = metadata
) {
    companion object {
        fun of(
            accountId: String,
            amount: Double,
            eventDate: LocalDateTime,
            metadata: EventMetadata
        ): MoneyWithdrawnEvent {
            return MoneyWithdrawnEvent(
                accountId = accountId,
                amount = amount,
                eventDate = eventDate,
                metadata = metadata
            )
        }
    }
}